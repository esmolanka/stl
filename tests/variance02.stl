
type Functor (±f : Type -> Type) =
  { map : forall a b. (a -> b) -> f a -> f b }

type Fix (f : Type -> Type) = f Fix

type Algebra (±f : Type -> Type) (±a : Type) =
  f { r : Fix f, q : a } -> a

type Paramorphism =
  forall (f : Type -> Type) (a : Type).
  Functor f -> Algebra f a -> Fix f  -> a

#eval Paramorphism

