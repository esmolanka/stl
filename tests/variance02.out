tests/variance02.types:14:1-14:19:
  Normalised:
    ∀ (f : Type -> Type). ∀ a.
      (∃ (ρ : Row). {map : ∀ a. ∀ b. (a -> b) -> (f a) -> f b | ρ})
      -> ((f (∃ (ρ : Row). {r : μ Fix. f Fix, q : a | ρ})) -> a)
      -> (μ Fix. f Fix) -> a
  Kind:
    Type